### PR TITLE
rishabh's submission

### PR DESCRIPTION
**Pull request by rishabh-ranjan**
val loss: __3.417__
wandb: <https://api.wandb.ai/links/rishabh-ranjan/93c9jode>

# description

After some iterative development, my data processing pipeline looks as follows:
1. __Extract text from WARC records.__
    I used `BeautifulSoup4` with the `lxml` parser for efficiency.
    I filter WARC response records with `Content-type == "text/html"`, ignoring records where `BeautifulSoup4` raises `MarkupResemblesLocatorWarning` or `XMLParsedAsHTMLWarning`.
2. __Identify exact line duplicates across all 5000 files.__
    My final implementation ended up being particularly efficient, completing in under __20__ minutes on a __288__ core machine for all __5000__ files.
    I allocate a 4G buffer with 1 byte for every possible 32-bit uint.
    I hashed each line into a 32-bit uint and increment the count of the hash in the buffer, but not if it would lead to uint8 overflow.
    I use `multiprocessing.shared_memory.SharedMemory` to let parallel python processes operate on the same hash count buffer.
    While race conditions are possible, (1) they are extremely rare, and (2) they would only lead to minor undercounting, which we are not sensitive to.
3. __Filter out duplicate lines.__
    Step 2 gives a hash count buffer (with counts capped at 255, but that is more than enough). Once loaded into memory, this allows efficiently detecting if a line is a duplicate (a hashing operation followed by an array indexing operation). For further efficiency, I hash the line as a python `bytes` object to delegate UTF-8 decoding only for the non-duplicate lines. Since, deduplication is so efficient it is the very first filter in my pipeline. Simply removing lines with hash count __>= 2__ removes around __86%__ of the extracted text bytes. Looking at the lines removed, I relaxed exact line deduplication a bit by only removing lines with hash count __>= 5__. This removes around __70%__ of the bytes. Interestingly, the hash count roughly follows a power law:  
    image  
    
    Another thing to note is that there is non-negligible chance of hash collisions. Out of the __~4.3e9__ buckets, __~2.5e9__ were filled.
4. __Remove toxic lines with Dolma FastText classifier with threshold 0.4 (following Dolma).__
5. __Remove NSFW lines with Dolma FastText classifier with threshold 0.4 (following Dolma).__
6. __Retain only English text with threshold 0.5 at the document level (following Dolma).__
7. __Apply some easy-to-implement quality rules from Gopher, and the C4 NoPunc rule (following Dolma).__
8. __Some miscellaneous tweaks such as removing consecutive duplicate lines, stripping whitespace, etc.__

This results in roughly __10B__ tokens. The given hyperparameters train over roughly __26B__ tokens, so this gives around 2.6 repetitions.

### CODE REVIEW

**Comment by nelson-liu** [comment]
thanks!!
